unversioned_cask_checker: check installer artifacts

bevanjkay opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR adds installer paths to the `unversioned_cask_checker` method, which allows the `:extract_plist` livecheck method to look inside of installer paths for `info.plist` files. This will be used fairly rarely but will enable a livecheck to be added to the PR - https://github.com/Homebrew/homebrew-cask/pull/128899


I'm seeing an error with typecheck locally, that I'm not sure how to resolve as this is not the case practically;

```
unversioned_cask_checker.rb:102: This code is unreachable https://srb.help/7006
     102 |          source = artifact.is_a?(Cask::Artifact::Installer) ? artifact.path : artifact.source.basename
                                                                         ^^^^^^^^^^^^^
    unversioned_cask_checker.rb:102: This condition was always falsy (T::Boolean)
     102 |          source = artifact.is_a?(Cask::Artifact::Installer) ? artifact.path : artifact.source.basename
```

Bo98 commented on Library/Homebrew/unversioned_cask_checker.rb:
```diff
@@ -36,6 +36,11 @@ def qlplugins
       @qlplugins ||= @cask.artifacts.select { |a| a.is_a?(Cask::Artifact::Qlplugin) }
     end
 
+    sig { returns(T::Array[Cask::Artifact::Qlplugin]) }
```
```suggestion
    sig { returns(T::Array[Cask::Artifact::Installer]) }
```

bevanjkay replied:
Thanks @Bo98 

Just gave myself away as a web dev with little typed language experience 😂

MikeMcQuaid reviewed (commented):
Looks good to me so far!

Would like to see this lines covered by at least unit tests, some other cask folks and/or @Bo98 👍🏻 this before merging.

reitermarkus commented on Library/Homebrew/unversioned_cask_checker.rb:
```diff
@@ -88,8 +98,9 @@ def all_versions
-        info_plist_paths = apps.concat(qlplugins).flat_map do |artifact|
-          top_level_info_plists(Pathname.glob(dir/"**"/artifact.source.basename/"Contents"/"Info.plist")).sort
+        info_plist_paths = apps.concat(qlplugins, installers).flat_map do |artifact|
+          source = artifact.is_a?(Cask::Artifact::Installer) ? artifact.path : artifact.source.basename
```
Why do installers need to use `path` here?

bevanjkay replied:
Most Cask::Artifact inherit `#source` from #Relocatable. However, installer does is not a child of Relocatable so does have `#source` but has `#path` instead.

SMillerDev replied:
Yeah, same in the signing audit

Moulick commented:
> Looks good to me so far!
> 
> Would like to see this lines covered by at least unit tests, some other cask folks and/or @Bo98 👍🏻 this before merging.

One cask that needs this is https://github.com/Homebrew/homebrew-cask-drivers/pull/3008

Bo98 commented on Library/Homebrew/unversioned_cask_checker.rb:
```diff
@@ -46,6 +51,11 @@ def single_app_cask?
       apps.count == 1
     end
 
+    sig { returns(T::Boolean) }
+    def single_installer_cask?
+      installers.count == 1
+    end
```
This doesn't appear to be used?

bevanjkay replied:
@Bo98 I was simply duplicating the logic that was previously in the class, it may not be necessary.

MikeMcQuaid replied:
@bevanjkay Please remove unless you know it's necessary.

Bo98 commented on Library/Homebrew/unversioned_cask_checker.rb:
```diff
@@ -88,8 +98,9 @@ def all_versions
 
-        info_plist_paths = apps.concat(qlplugins).flat_map do |artifact|
-          top_level_info_plists(Pathname.glob(dir/"**"/artifact.source.basename/"Contents"/"Info.plist")).sort
+        info_plist_paths = apps.concat(qlplugins, installers).flat_map do |artifact|
+          source = artifact.is_a?(Cask::Artifact::Installer) ? artifact.path : artifact.source.basename
+          top_level_info_plists(Pathname.glob(dir/"**"/source/"Contents"/"Info.plist")).sort
         end
```
Is there a reason for this being different than line 151-153 below?

bevanjkay replied:
No, the logic likely should be the same. I'm not sure how often we are using that method, the only place I could find it was as a part of the `bump-unversioned-casks` dev-cmd.

Moulick commented:
Gentle nudge as I want to get https://github.com/Homebrew/homebrew-cask-drivers/pull/3008 working and closed :)

MikeMcQuaid reviewed (approved):
Fine with me! Ideally would have a unit test and some dead code cleaned up if necessary.

Moulick commented:
Gentle nudge :)

bevanjkay commented:
I have added a test that should confirm that this works with an `installer` artifact.
I have little experience with writing tests, so I hope that this suffices.

To allow the test to be completed, I needed an artifact to check for a contained `info.plist` file. I modified the existing caffeine fixture by adding a minimal `info.plist` file. Please let me know if it would be better to create a separate fixture for this test instead.

bevanjkay commented:
It seems that the test doesn't work on Linux? I'm not sure how to proceed, would love some direction here.

MikeMcQuaid commented:
> It seems that the test doesn't work on Linux? I'm not sure how to proceed, would love some direction here.

Looks like a Sorbet issue here where the wrong argument is being passed?

bevanjkay commented:
The test is simulating the `livecheck` command on a cask fixture, however this will not work on Linux as the strategy uses system commands that are only found on MacOS - https://github.com/Homebrew/brew/blob/96574c831f9ed1f3fa281f4724719d8cb2424ece/Library/Homebrew/unversioned_cask_checker.rb#L83

Could this be why it is failing on CI, but I can't replicate it locally?

aaronliu0130 commented:
Maybe restrict feature to macOS only? or find a linux alternative?

bevanjkay commented:
@aaronliu0130 The whole cask system should be limited to Mac only, but I'm not sure how that integrates with CI testing.

MikeMcQuaid commented:
@bevanjkay Can this command perhaps be mocked out instead? It'll be slower to call it directly anyway.

MikeMcQuaid commented:
Thanks again @bevanjkay!
